### PR TITLE
WT-6751 Assert if new value is the same as old value for a history store update

### PR DESCRIPTION
**Pull request by hbokhari**

### CODE REVIEW

File: src/history/hs_rec.c [hunk 1]
```diff
@@ -213,8 +229,22 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
...
+
+#ifdef HAVE_DIAGNOSTIC
+            if (cmp == 0) {
+                WT_ERR(cursor->get_value(cursor, &hs_stop_durable_ts_diag, &durable_timestamp_diag,
+                  &upd_type_full_diag, existing_val));
+                WT_ERR(__wt_compare(session, NULL, existing_val, hs_value, &cmp));
+                if (cmp == 0)
+                    WT_ASSERT(session,
+                      start_time_point->txnid == WT_TXN_NONE ||
+                        start_time_point->txnid != hs_cbt->upd_value->tw.start_txn);
```
Comment (quchenhao) [inline src/history/hs_rec.c:241]: It's also possible we have the same txids but different timestamps so I think we should check timestamps as well.

File: src/history/hs_rec.c [hunk 2]
```diff
@@ -213,8 +229,23 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
...
             WT_ERR(__wt_compare(session, NULL, hs_key, key, &cmp));
+
+#ifdef HAVE_DIAGNOSTIC
+            if (cmp == 0) {
+                WT_ERR(cursor->get_value(cursor, &hs_stop_durable_ts_diag, &durable_timestamp_diag,
+                  &upd_type_full_diag, existing_val));
+                WT_ERR(__wt_compare(session, NULL, existing_val, hs_value, &cmp));
+                if (cmp == 0)
+                    WT_ASSERT(session,
+                      start_time_point->txnid == WT_TXN_NONE ||
```
Comment (quchenhao) [inline src/history/hs_rec.c:240]: I'm thinking the necessity of checking start_time_point->txnid == WT_TXN_NONE. If both updates have txnid 0 and their ts are the same, they will be squashed as well. So I think it is enough to just check the txnids and timestamps are different.

File: src/history/hs_rec.c [hunk 3]
```diff
@@ -213,8 +229,23 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
...
+                        start_time_point->txnid != hs_cbt->upd_value->tw.start_txn ||
+                        start_time_point->ts != hs_cbt->upd_value->tw.start_ts);
+                counter = hs_counter + 1;
+            }
```
Comment (quchenhao) [inline src/history/hs_rec.c:244]: I think it is worth running a full set of stress tests and mongodb debug tests to make sure this doesn't crash on any corner cases.

Comment (hbokhari) [reply src/history/hs_rec.c:244]: Agree. I'll start patch builds and add links to the ticket.

Comment (hbokhari) [reply src/history/hs_rec.c:244]: @quchenhao @bvpvamsikrishna - The MongoDB patch looks good.

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Pull request merged**